src/VisualStudio/Core/Def/Progression/GraphQueryManager.cs

```diff
@@ -21,162 +25,129 @@ namespace Microsoft.VisualStudio.LanguageServices.Implementation.Progression
         private readonly object _gate = new();
-        private readonly List<ValueTuple<WeakReference<IGraphContext>, List<IGraphQuery>>> _trackedQueries = new();
+        private ImmutableArray<(WeakReference<IGraphContext> context, ImmutableArray<IGraphQuery> queries)> _trackedQueries = ImmutableArray<(WeakReference<IGraphContext>, ImmutableArray<IGraphQuery>)>.Empty;
 
-        // We update all of our tracked queries when this delay elapses.
-        private ResettableDelay? _delay;
+        private readonly AsyncBatchingWorkQueue _updateQueue;
 
-        internal GraphQueryManager(Workspace workspace, IAsynchronousOperationListener asyncListener)
+        internal GraphQueryManager(
+            Workspace workspace,
+            IThreadingContext threadingContext,
+            IAsynchronousOperationListener asyncListener)
         {
             _workspace = workspace;
-            _asyncListener = asyncListener;
-        }
-
-        internal void AddQueries(IGraphContext context, List<IGraphQuery> graphQueries)
-        {
-            var asyncToken = _asyncListener.BeginAsyncOperation("GraphQueryManager.AddQueries");
-
-            var solution = _workspace.CurrentSolution;
-
-            var populateTask = PopulateContextGraphAsync(solution, graphQueries, context);
-
-            // We want to ensure that no matter what happens, this initial context is completed
-            var task = populateTask.SafeContinueWith(
-                _ => context.OnCompleted(), context.CancelToken, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
-
-            if (context.TrackChanges)
-            {
-                task = task.SafeContinueWith(
-                    _ => TrackChangesAfterFirstPopulate(graphQueries, context, solution),
-                    context.CancelToken, TaskContinuationOptions.None, TaskScheduler.Default);
-            }
-
-            task.CompletesAsyncOperation(asyncToken);
-        }
 
-        private void TrackChangesAfterFirstPopulate(List<IGraphQuery> graphQueries, IGraphContext context, Solution solution)
-        {
-            var workspace = solution.Workspace;
-            var contextWeakReference = new WeakReference<IGraphContext>(context);
+            // Update any existing live/tracking queries 1.5 seconds after every workspace changes.
+            _updateQueue = new AsyncBatchingWorkQueue(
+                DelayTimeSpan.Idle,
+                UpdateExistingQueriesAsync,
+                asyncListener,
+                threadingContext.DisposalToken);
 
-            lock (_gate)
-            {
-                if (_trackedQueries.IsEmpty())
-                {
-                    _workspace.WorkspaceChanged += OnWorkspaceChanged;
-                }
-
-                _trackedQueries.Add(ValueTuple.Create(contextWeakReference, graphQueries));
-            }
-
-            EnqueueUpdateIfSolutionIsStale(solution);
+            _workspace.WorkspaceChanged += (_, _) => _updateQueue.AddWork();
```

Comment:
It looks like before we were trying to avoid having a WorkspaceChanged event subscribed if we didn't need one -- any sense if that optimization even mattered in practice? I'd imagine there was _always_ at least one thing listening, right?

Reply:
correct. i'll add a comment about this.